File: app/src/main/java/com/example/task2/ui/benchmark/BenchmarksFragment.java
package com.example.task2.ui.benchmark;

import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.CompoundButton;
import android.widget.PopupWindow;
import android.widget.TextView;
import android.widget.ToggleButton;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.example.task2.R;
import com.google.android.material.textfield.TextInputEditText;

public class BenchmarksFragment extends Fragment implements View.OnFocusChangeListener, TextWatcher, CompoundButton.OnCheckedChangeListener {
    private static final String ARG_BENCHMARK_TYPE = "benchmarkType";
    private final BenchmarksAdapter adapter = new BenchmarksAdapter();
    private TextInputEditText textInputEditText;
    private ToggleButton buttonStopStart;
    private PopupWindow errorPopup;
    private BenchmarksViewModel viewModel;

    public static BenchmarksFragment newInstance(int benchmarkType) {
        Bundle args = new Bundle();
        args.putInt(ARG_BENCHMARK_TYPE, benchmarkType);
        BenchmarksFragment fragment = new BenchmarksFragment();
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        Bundle args = getArguments();
        int benchmarkType = args.getInt(ARG_BENCHMARK_TYPE);

        BenchmarksViewModelFactory factory = new BenchmarksViewModelFactory(benchmarkType);
        viewModel = new ViewModelProvider(this, factory).get(BenchmarksViewModel.class);
        viewModel.onCreate();
    }

    @Override
    public View onCreateView(
            LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState
    ) {
        return inflater.inflate(R.layout.fragment_benchmark, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        final RecyclerView recyclerView = view.findViewById(R.id.rv);
        final GridLayoutManager layoutManager = new GridLayoutManager(
                getActivity(), viewModel.getNumberOfColumns()
        );
        recyclerView.setLayoutManager(layoutManager);
        recyclerView.setAdapter(adapter);

        viewModel.getCellOperationsLiveData().observe(getViewLifecycleOwner(), adapter::submitList);
        viewModel.getAllTasksCompletedLiveData().observe(getViewLifecycleOwner(), allTasksCompleted -> {
            buttonStopStart.setOnCheckedChangeListener(null);
            buttonStopStart.setChecked(allTasksCompleted);
            buttonStopStart.setOnCheckedChangeListener(this);
        });
        viewModel.getValidNumberLiveData().observe(getViewLifecycleOwner(), errorMessage -> {
            if (errorMessage == null) {
                buttonStopStart.setEnabled(true);
                textInputEditText.setBackgroundResource(R.drawable.input_bg2);
                if (errorPopup != null) {
                    errorPopup.dismiss();
                }
            } else {
                buttonStopStart.setEnabled(false);
                showError(errorMessage, R.drawable.input_bg_error);
            }
        });

        textInputEditText = view.findViewById(R.id.editText);
        textInputEditText.setOnFocusChangeListener(this);
        textInputEditText.addTextChangedListener(this);

        buttonStopStart = view.findViewById(R.id.btnStopStart);
        buttonStopStart.setOnCheckedChangeListener(this);
        buttonStopStart.setEnabled(false);
    }

    public void onFocusChange(View v, boolean hasFocus) {
        if (!hasFocus) {
            textInputEditText.clearFocus();
            textInputEditText.setBackgroundResource(R.drawable.input_bg2);
            if (errorPopup != null) {
                errorPopup.dismiss();
            }
        }
    }

    @Override
    public void beforeTextChanged(CharSequence s, int start, int count, int after) {

    }

    @Override
    public void onTextChanged(CharSequence s, int start, int before, int count) {
        final String input = s.toString().trim();

        viewModel.validateNumber(input);
    }

    @Override
    public void afterTextChanged(Editable s) {

    }

    @Override
    public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
        viewModel.onButtonClicked(textInputEditText.getText().toString().trim());
    }

    private void showError(int errorMessage, int inputBg) {
        final View errorView;
        final TextView errorText;
        if (errorPopup == null) {
            errorView = LayoutInflater.from(getContext()).inflate(R.layout.view_error, (ViewGroup) getView(), false);
            errorText = errorView.findViewById(R.id.errorText);
            errorPopup = new PopupWindow(errorView, ViewGroup.LayoutParams.WRAP_CONTENT, ViewGroup.LayoutParams.WRAP_CONTENT);
        } else {
            errorView = errorPopup.getContentView();
            errorText = errorPopup.getContentView().findViewById(R.id.errorText);
        }
        final int Y = 30;

        errorText.setText(errorMessage);
        textInputEditText.setBackgroundResource(inputBg);

        errorPopup.setWidth(ViewGroup.LayoutParams.WRAP_CONTENT);
        errorPopup.setHeight(ViewGroup.LayoutParams.WRAP_CONTENT);
        errorPopup.setContentView(errorView);

        errorView.measure(View.MeasureSpec.UNSPECIFIED, View.MeasureSpec.UNSPECIFIED);
        final int errorPopupWidth = errorView.getMeasuredWidth();
        final int x = textInputEditText.getWidth() / 2 - errorPopupWidth / 2;

        errorPopup.showAsDropDown(textInputEditText, x, Y);
    }
}

File: app/src/main/java/com/example/task2/ui/benchmark/BenchmarksFragment.kt
package com.example.task2.ui.benchmark

import android.app.Dialog
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.LayoutInflater
import android.view.View
import android.view.View.OnFocusChangeListener
import android.view.ViewGroup
import android.widget.CompoundButton
import android.widget.PopupWindow
import android.widget.TextView
import android.widget.ToggleButton
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.example.task2.R
import com.example.task2.ui.SpacesItemDecoration
import com.google.android.material.textfield.TextInputEditText
import kotlinx.coroutines.launch

class BenchmarksFragment : Fragment(), OnFocusChangeListener, TextWatcher,
        CompoundButton.OnCheckedChangeListener, View.OnLongClickListener {

    companion object {
        const val ARG_BENCHMARK_TYPE = "benchmarkType"

        fun newInstance(benchmarkType: Int): BenchmarksFragment {
            return BenchmarksFragment().apply {
                arguments = Bundle().apply {
                    putInt(ARG_BENCHMARK_TYPE, benchmarkType)
                }
            }
        }
    }

    private val adapter = BenchmarksAdapter()
    private lateinit var textInputEditText: TextInputEditText
    private lateinit var buttonStopStart: ToggleButton
    private var errorPopup: PopupWindow? = null
    private lateinit var viewModel: BenchmarksViewModel

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        val factory = BenchmarksViewModelFactory(requireArguments().getInt(ARG_BENCHMARK_TYPE))
        viewModel = ViewModelProvider(this, factory)[BenchmarksViewModel::class.java]
        viewModel.onCreate()
    }

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?
    ): View? = inflater.inflate(R.layout.fragment_benchmark, container, false)

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        view.findViewById<TextView>(R.id.textView)?.setOnLongClickListener(this)

        val recyclerView = view.findViewById<RecyclerView>(R.id.rv)
        recyclerView.layoutManager = GridLayoutManager(activity, viewModel.getNumberOfColumns())
        recyclerView.adapter = adapter
        recyclerView.addItemDecoration(SpacesItemDecoration(viewModel.getNumberOfColumns(), resources.getDimensionPixelSize(R.dimen.card_margin)))

        viewModel.getCellOperationsLiveData().observe(viewLifecycleOwner) { adapter.submitList(it) }
        viewModel.getAllTasksCompletedLiveData().observe(viewLifecycleOwner) { allTasksCompleted ->
            buttonStopStart.setOnCheckedChangeListener(null)
            buttonStopStart.isChecked = allTasksCompleted
            buttonStopStart.setOnCheckedChangeListener(this)
        }
        viewModel.getValidNumberLiveData().observe(viewLifecycleOwner) { errorMessage: Int? ->
            if (errorMessage == null) {
                buttonStopStart.isEnabled = true
                textInputEditText.setBackgroundResource(R.drawable.input_bg2)
                if (errorPopup != null) {
                    errorPopup?.dismiss()
                }
            } else {
                buttonStopStart.isEnabled = false
                showError(errorMessage, R.drawable.input_bg_error)
            }
        }

        textInputEditText = view.findViewById(R.id.editText)
        textInputEditText.onFocusChangeListener = this
        textInputEditText.addTextChangedListener(this)

        buttonStopStart = view.findViewById(R.id.btnStopStart)
        buttonStopStart.setOnCheckedChangeListener(this)
        buttonStopStart.isEnabled = false
    }

    override fun onFocusChange(v: View, hasFocus: Boolean) {
        if (!hasFocus) {
            textInputEditText.clearFocus()
            textInputEditText.setBackgroundResource(R.drawable.input_bg2)
            errorPopup?.dismiss()
        }
    }

    override fun beforeTextChanged(s: CharSequence, start: Int, count: Int, after: Int) {}
    override fun onTextChanged(s: CharSequence, start: Int, before: Int, count: Int) {
        viewModel.validateNumber(s.toString().trim())
    }

    override fun afterTextChanged(s: Editable) {}

    override fun onCheckedChanged(buttonView: CompoundButton, isChecked: Boolean) {
        viewModel.onButtonClicked(textInputEditText.text.toString().trim())
    }

    private fun showError(errorMessage: Int, inputBg: Int) {
        lateinit var errorView: View
        lateinit var errorText: TextView
        if (errorPopup == null) {
            errorView =
                    LayoutInflater.from(context).inflate(R.layout.view_error, view as ViewGroup?, false)
            errorText = errorView.findViewById(R.id.errorText)
            errorPopup = PopupWindow(
                    errorView,
                    ViewGroup.LayoutParams.WRAP_CONTENT,
                    ViewGroup.LayoutParams.WRAP_CONTENT
            )
        } else {
            errorPopup?.let {
                errorView = it.contentView
                errorText = it.contentView.findViewById(R.id.errorText)
            }
        }

        errorText.setText(errorMessage)
        textInputEditText.setBackgroundResource(inputBg)

        errorPopup?.apply {
            width = ViewGroup.LayoutParams.WRAP_CONTENT
            height = ViewGroup.LayoutParams.WRAP_CONTENT
            contentView = errorView

            val y = 30
            errorView.measure(View.MeasureSpec.UNSPECIFIED, View.MeasureSpec.UNSPECIFIED)
            val errorPopupWidth = errorView.measuredWidth
            val x = textInputEditText.width / 2 - errorPopupWidth / 2
            showAsDropDown(textInputEditText, x, y)
        }
    }

    override fun onLongClick(p0: View?): Boolean {
        lifecycleScope.launch {
            val last21Result = viewModel.getLast21Results().reversed()
            val textToShow = StringBuilder()
            for ((index, result) in last21Result.withIndex()) {
                val action = result.action.replace("\n", "")
                if (index == 0) {
                    textToShow.append("Input: ${result.input}\n")
                }
                textToShow.append("$action ${result.type} ${result.time} n/a\n")
            }

            val inflater = LayoutInflater.from(requireContext())
            val layout = inflater.inflate(R.layout.toast_bd, null)

            val customToastText = layout.findViewById<TextView>(R.id.customToastText)
            customToastText.text = textToShow.toString()

            val dialog = Dialog(requireContext())
            dialog.setContentView(layout)
            dialog.window?.setBackgroundDrawableResource(android.R.color.transparent)


            val displayMetrics = context?.resources?.displayMetrics
            val width = (displayMetrics?.widthPixels?.times(0.9))?.toInt()
            width?.let {
                dialog.window?.setLayout(
                        it,
                        ViewGroup.LayoutParams.WRAP_CONTENT
                )
            }
            dialog.show()
        }
        return true
    }
}
